plugin/db/mysql/binlog.go

```diff
@@ -0,0 +1,265 @@
+	UpdateRowsEvent
+	// WriteRowsEvent is the binlog event for INSERT.
+	WriteRowsEvent
+	// QueryEvent is the binlog event for QUERY.
+	// The thread ID is parsed from QUERY events.
+	QueryEvent
+)
+
+var (
+	regexpDeleteRow = regexp.MustCompile("DELETE FROM `(.+)`\\.`(.+)`")
+	regexpUpdateRow = regexp.MustCompile("UPDATE `(.+)`\\.`(.+)`")
+	regexpWriteRow  = regexp.MustCompile("INSERT INTO `(.+)`\\.`(.+)`")
+	regexpThreadID  = regexp.MustCompile(`thread_id=(\d+)`)
+)
+
+func (t binlogEventType) String() string {
+	switch t {
+	case DeleteRowsEvent:
+		return "DELETE"
+	case UpdateRowsEvent:
+		return "UPDATE"
+	case WriteRowsEvent:
+		return "INSERT"
+	default:
+		return "UNKNOWN"
+	}
+}
+
+func (t binlogEventType) Regexp() *regexp.Regexp {
+	switch t {
+	case DeleteRowsEvent:
+		return regexpDeleteRow
+	case UpdateRowsEvent:
+		return regexpUpdateRow
+	case WriteRowsEvent:
+		return regexpWriteRow
+	default:
+		return nil
+	}
+}
+
+func (t binlogEventType) MinBlockLen() int {
+	switch t {
+	case DeleteRowsEvent:
+		return 3
+	case UpdateRowsEvent:
+		return 5
+	case WriteRowsEvent:
+		return 3
+	default:
+		return -1
+	}
+}
+
+func (t binlogEventType) ParseDMLPayload(block []string) (beforeValue []string, afterValue []string, err error) {
+	block = block[1:]
+	switch t {
+	case DeleteRowsEvent:
+		// Example block:
+		// ### DELETE FROM `database`.`table`
+		// ### WHERE
+		// ###   @1=x
+		//       ...
+		where, err := parseBinlogEventPayloadBlock(block, "WHERE")
+		if err != nil {
+			return nil, nil, errors.Wrap(err, "failed to parse the WHERE payload section of the DELETE event")
+		}
+		return where, nil, nil
+	case UpdateRowsEvent:
+		// Example block:
+		// ### UPDATE `database`.`table`
+		// ### WHERE
+		// ###   @1=x
+		//       ...
+		// ### SET
+		// ###   @1=y
+		// 	     ...
+		if len(block)%2 != 0 {
+			return nil, nil, errors.Errorf("invalid UPDATE event block, WHERE clause length != SET clause length: %q", strings.Join(block, "\n"))
+		}
+		where, err := parseBinlogEventPayloadBlock(block[:len(block)/2], "WHERE")
+		if err != nil {
+			return nil, nil, errors.Wrap(err, "failed to parse the WHERE payload section of the UPDATE event")
+		}
+		block = block[len(block)/2:]
+		set, err := parseBinlogEventPayloadBlock(block, "SET")
+		if err != nil {
+			return nil, nil, errors.Wrap(err, "failed to parse the SET payload section of the UPDATE event")
+		}
+		return where, set, nil
+	case WriteRowsEvent:
+		// Example block:
+		// ## INSERT INTO `database`.`table`
+		// ### SET
+		// ###   @1=x
+		//       ...
+		set, err := parseBinlogEventPayloadBlock(block, "SET")
+		if err != nil {
+			return nil, nil, errors.Wrap(err, "failed to parse the SET payload section of the INSERT event")
+		}
+		return nil, set, nil
+	default:
+		return nil, nil, errors.Errorf("invalid DML binlog event type %s", t.String())
+	}
+}
+
+type binlogEvent struct {
+	Type       binlogEventType
+	Database   string
+	Table      string
+	DataBefore [][]string
```

Comment:
We shouldn’t need DataBefore/After in the binlogEvent format. That’s why we need an SQL parser. Or was it performance related?

Reply:
The binlog text is a customized format from mysqlbinlog. We have to parse and store it.